File: frontend/lib/api/productManagement.ts
/**
 * TypeScript API Client for Phase 56 — Unified Product Lifecycle, Product Management & Continuous Delivery Intelligence Platform.
 */

export interface ProductItem {
  id: string;
  name: string;
  type: string;
  description?: string;
  target_market?: string;
  owner: string;
  team?: string;
  lifecycle_stage: string;
  health: string;
  status: string;
  version: string;
  workspace_id?: string;
  created_at: string;
  updated_at: string;
}

export interface ProductObjective {
  id: string;
  product_id: string;
  name: string;
  metric: string;
  baseline: number;
  target: number;
  time_window: string;
  owner: string;
  confidence: number;
  status: string;
}

export interface ProductMetric {
  id: string;
  product_id: string;
  name: string;
  definition: string;
  formula?: string;
  source: string;
  owner: string;
  current_value: number;
  target_value?: number;
  is_north_star: boolean;
}

export interface ProductFeedback {
  id: string;
  product_id: string;
  source: string;
  feedback_type: string;
  customer_segment: string;
  content: string;
  sentiment_score?: number;
  revenue_impact_usd?: number;
  created_at: string;
}

export interface FeedbackTheme {
  theme: string;
  frequency: number;
  sentiment_avg: number;
  revenue_impact_total: number;
  urgency_level: string;
}

export interface ProductRequirement {
  id: string;
  product_id: string;
  title: string;
  description: string;
  category: string;
  priority: string;
  acceptance_criteria: string[];
  source: string;
  problem_id?: string;
  objective_id?: string;
  is_traceable: boolean;
  version: number;
}

export interface TraceabilityMatrix {
  product_id: string;
  requirements: ProductRequirement[];
  orphaned_requirements: ProductRequirement[];
  traceability_score: number;
  unlinked_backlog_items: any[];
}

export interface ProductEpic {
  id: string;
  product_id: string;
  title: string;
  objective: string;
  problem_statement?: string;
  status: string;
}

export interface ProductFeature {
  id: string;
  product_id: string;
  epic_id?: string;
  requirement_id?: string;
  title: string;
  description?: string;
  acceptance_criteria: string[];
  status: string;
}

export interface ProductBacklogItem {
  id: string;
  product_id: string;
  epic_id?: string;
  feature_id?: string;
  requirement_id?: string;
  title: string;
  type: string;
  story_points: number;
  status: string;
  prioritization_score?: number;
  prioritization_framework?: string;
}

export interface ProductRoadmap {
  id: string;
  product_id: string;
  title: string;
  scenario: string;
  items: ProductRoadmapItem[];
}

export interface ProductRoadmapItem {
  id: string;
  roadmap_id: string;
  title: string;
  horizon: string;
  feature_id?: string;
  confidence: number;
  objective_link?: string;
}

export interface ProductSprint {
  id: string;
  product_id: string;
  name: string;
  status: string;
  capacity_points: number;
  start_date?: string;
  end_date?: string;
}

export interface ProductRelease {
  id: string;
  product_id: string;
  version: string;
  readiness_status: string;
  critical_defects: number;
  qa_passed: boolean;
  security_reviewed: boolean;
  performance_benchmarked: boolean;
  rollback_tested: boolean;
  launch_readiness_score: number;
  blocking_reasons: string[];
}

export interface ProductFeatureFlag {
  id: string;
  product_id: string;
  key: string;
  state: string;
  rollout_percentage: number;
  description?: string;
}

export interface ProductHealthDetail {
  product_id: string;
  composite_score: number;
  status: string;
  factors: {
    adoption: number;
    satisfaction: number;
    reliability: number;
    velocity: number;
    revenue: number;
    security: number;
  };
}

export interface ProductDashboardData {
  product: ProductItem;
  vision: any;
  objectives: ProductObjective[];
  metrics: ProductMetric[];
  feedback_count: number;
  requirements_count: number;
  backlog_count: number;
  roadmaps_count: number;
  sprints_count: number;
  releases_count: number;
  health: ProductHealthDetail;
}

const API_BASE = process.env.NEXT_PUBLIC_API_URL || "";

export const productApi = {
  async getPortfolio(): Promise<{ total_products: number; active_products: number; products: ProductItem[] }> {
    const res = await fetch(`${API_BASE}/api/v1/products/portfolio`);
    if (!res.ok) throw new Error("Failed to load product portfolio");
    const json = await res.json();
    return json.data;
  },

  async listProducts(stage?: string): Promise<ProductItem[]> {
    const url = stage ? `${API_BASE}/api/v1/products?stage=${stage}` : `${API_BASE}/api/v1/products`;
    const res = await fetch(url);
    if (!res.ok) throw new Error("Failed to list products");
    const json = await res.json();
    return json.data;
  },

  async createProduct(payload: {
    name: string;
    type?: string;
    description?: string;
    target_market?: string;
    owner?: string;
    team?: string;
  }): Promise<ProductItem> {
    const res = await fetch(`${API_BASE}/api/v1/products`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });
    if (!res.ok) throw new Error("Failed to create product");
    const json = await res.json();
    return json.data;
  },

  async getProductDashboard(id: string): Promise<ProductDashboardData> {
    const res = await fetch(`${API_BASE}/api/v1/products/${id}`);
    if (!res.ok) throw new Error(`Failed to load product ${id}`);
    const json = await res.json();
    return json.data;
  },

  async transitionLifecycle(id: string, new_stage: string, rationale?: string): Promise<ProductItem> {
    const res = await fetch(`${API_BASE}/api/v1/products/${id}/lifecycle`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ new_stage, rationale }),
    });
    if (!res.ok) throw new Error("Failed to transition lifecycle stage");
    const json = await res.json();
    return json.data;
  },

  async setVision(id: string, payload: any): Promise<any> {
    const res = await fetch(`${API_BASE}/api/v1/products/${id}/vision`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });
    if (!res.ok) throw new Error("Failed to set vision");
    const json = await res.json();
    return json.data;
  },

  async getTraceability(id: string): Promise<TraceabilityMatrix> {
    const res = await fetch(`${API_BASE}/api/v1/products/${id}/traceability`);
    if (!res.ok) throw new Error("Failed to load traceability matrix");
    const json = await res.json();
    return json.data;
  },

  async createRequirement(id: string, payload: any): Promise<ProductRequirement> {
    const res = await fetch(`${API_BASE}/api/v1/products/${id}/requirements`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });
    if (!res.ok) throw new Error("Failed to create requirement");
    const json = await res.json();
    return json.data;
  },

  async getBacklog(id: string): Promise<{ epics: ProductEpic[]; features: ProductFeature[]; items: ProductBacklogItem[] }> {
    const res = await fetch(`${API_BASE}/api/v1/products/${id}/backlog`);
    if (!res.ok) throw new Error("Failed to load backlog");
    const json = await res.json();
    return json;
  },

  async scoreBacklogItem(productId: string, itemId: string, framework: string, inputs: Record<string, number>): Promise<any> {
    const res = await fetch(`${API_BASE}/api/v1/products/${productId}/backlog/score/${itemId}`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ framework, inputs }),
    });
    if (!res.ok) throw new Error("Failed to score backlog item");
    const json = await res.json();
    return json.data;
  },

  async evaluateReleaseReadiness(productId: string, releaseId: string, checks: any): Promise<any> {
    const res = await fetch(`${API_BASE}/api/v1/products/${productId}/releases/${releaseId}/readiness`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(checks),
    });
    if (!res.ok) throw new Error("Failed to evaluate release readiness");
    const json = await res.json();
    return json.data;
  },

  async queryCopilot(productId: string, query: string): Promise<any> {
    const res = await fetch(`${API_BASE}/api/v1/products/${productId}/copilot`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ query }),
    });
    if (!res.ok) throw new Error("Failed to query product copilot");
    const json = await res.json();
    return json.data;
  },
};

File: backend/app/api/v1/product_management.py
"""
REST API Router for Phase 56 — Unified Product Lifecycle, Product Management & Continuous Delivery Intelligence Platform.
"""

from typing import Dict, Any, List, Optional
from fastapi import APIRouter, HTTPException, Query, status

from backend.app.schemas.product_management import (
    ProductCreateRequest,
    LifecycleTransitionRequest,
    VisionSetRequest,
    ObjectiveCreateRequest,
    MetricRegisterRequest,
    FeedbackIngestRequest,
    RequirementCreateRequest,
    EpicCreateRequest,
    FeatureCreateRequest,
    BacklogItemCreateRequest,
    PrioritizationScoreRequest,
    RoadmapCreateRequest,
    RoadmapItemAddRequest,
    SprintCreateRequest,
    ReleaseCreateRequest,
    ReleaseReadinessEvaluateRequest,
    FeatureFlagCreateRequest,
    LaunchChecklistUpdateRequest,
    ExperimentCreateRequest,
    HealthSnapshotRequest,
    SunsetPlanCreateRequest,
    ProductCopilotRequest,
)
from backend.app.services.product_management.service import global_product_management_service

router = APIRouter(prefix="/products", tags=["Product Management & Lifecycle Platform"])


@router.get("/portfolio", summary="Get Portfolio Overview")
async def get_portfolio() -> Dict[str, Any]:
    portfolio = global_product_management_service.get_portfolio_overview()
    return {"status": "SUCCESS", "data": portfolio}


@router.get("", summary="List All Products")
async def list_products(
    stage: Optional[str] = Query(None),
    product_type: Optional[str] = Query(None),
) -> Dict[str, Any]:
    products = global_product_management_service.products.list_products(stage=stage, product_type=product_type)
    return {"status": "SUCCESS", "count": len(products), "data": products}


@router.post("", status_code=status.HTTP_201_CREATED, summary="Create New Product")
async def create_product(req: ProductCreateRequest) -> Dict[str, Any]:
    prod = global_product_management_service.products.create_product(
        name=req.name,
        type=req.type,
        description=req.description,
        target_market=req.target_market,
        owner=req.owner,
        team=req.team,
        workspace_id=req.workspace_id,
    )
    return {"status": "SUCCESS", "data": prod}


@router.get("/{id}", summary="Get Full Product Details & Subsystems")
async def get_product(id: str) -> Dict[str, Any]:
    try:
        data = global_product_management_service.get_product_dashboard(id)
        return {"status": "SUCCESS", "data": data}
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.post("/{id}/lifecycle", summary="Transition Product Lifecycle Stage")
async def transition_lifecycle(id: str, req: LifecycleTransitionRequest) -> Dict[str, Any]:
    try:
        updated = global_product_management_service.products.transition_lifecycle_stage(
            product_id=id,
            new_stage=req.new_stage,
            actor=req.actor,
            rationale=req.rationale,
        )
        return {"status": "SUCCESS", "data": updated}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/{id}/vision", summary="Define Product Vision & Strategy")
async def set_vision(id: str, req: VisionSetRequest) -> Dict[str, Any]:
    vision = global_product_management_service.vision_strategy.set_product_vision(
        product_id=id,
        target_customer=req.target_customer,
        problem_statement=req.problem_statement,
        desired_future_state=req.desired_future_state,
        value_proposition=req.value_proposition,
        differentiation=req.differentiation,
        strategic_alignment=req.strategic_alignment,
        success_definition=req.success_definition,
    )
    return {"status": "SUCCESS", "data": vision}


@router.get("/{id}/objectives", summary="List Product Objectives & OKRs")
async def list_objectives(id: str) -> Dict[str, Any]:
    objs = global_product_management_service.vision_strategy.list_objectives(id)
    return {"status": "SUCCESS", "count": len(objs), "data": objs}


@router.post("/{id}/objectives", status_code=status.HTTP_201_CREATED, summary="Create Product Objective")
async def create_objective(id: str, req: ObjectiveCreateRequest) -> Dict[str, Any]:
    obj = global_product_management_service.vision_strategy.create_objective(
        product_id=id,
        name=req.name,
        metric=req.metric,
        baseline=req.baseline,
        target=req.target,
        time_window=req.time_window,
        owner=req.owner,
        confidence=req.confidence,
    )
    return {"status": "SUCCESS", "data": obj}


@router.get("/{id}/metrics", summary="List Product Metrics Registry")
async def list_metrics(id: str) -> Dict[str, Any]:
    metrics = global_product_management_service.vision_strategy.list_metrics(id)
    return {"status": "SUCCESS", "count": len(metrics), "data": metrics}


@router.post("/{id}/metrics", status_code=status.HTTP_201_CREATED, summary="Register Metric")
async def register_metric(id: str, req: MetricRegisterRequest) -> Dict[str, Any]:
    met = global_product_management_service.vision_strategy.register_metric(
        product_id=id,
        name=req.name,
        definition=req.definition,
        formula=req.formula,
        source=req.source,
        owner=req.owner,
        current_value=req.current_value,
        target_value=req.target_value,
        is_north_star=req.is_north_star,
    )
    return {"status": "SUCCESS", "data": met}


@router.get("/{id}/feedback", summary="List Customer Feedback & Clusters")
async def list_feedback(id: str) -> Dict[str, Any]:
    fb = global_product_management_service.feedback_intelligence.list_feedback(id)
    themes = global_product_management_service.feedback_intelligence.cluster_feedback_into_themes(id)
    return {"status": "SUCCESS", "feedback_count": len(fb), "themes_count": len(themes), "feedback": fb, "themes": themes}


@router.post("/{id}/feedback", status_code=status.HTTP_201_CREATED, summary="Ingest Customer Feedback")
async def ingest_feedback(id: str, req: FeedbackIngestRequest) -> Dict[str, Any]:
    fb = global_product_management_service.feedback_intelligence.ingest_feedback(
        product_id=id,
        source=req.source,
        feedback_type=req.feedback_type,
        customer_segment=req.customer_segment,
        content=req.content,
        sentiment_score=req.sentiment_score,
        revenue_impact_usd=req.revenue_impact_usd,
    )
    return {"status": "SUCCESS", "data": fb}


@router.get("/{id}/traceability", summary="Get Complete Traceability Matrix & Orphan Detection")
async def get_traceability(id: str) -> Dict[str, Any]:
    matrix = global_product_management_service.requirements_traceability.get_traceability_matrix(id)
    return {"status": "SUCCESS", "data": matrix}


@router.post("/{id}/requirements", status_code=status.HTTP_201_CREATED, summary="Create PRD Requirement")
async def create_requirement(id: str, req: RequirementCreateRequest) -> Dict[str, Any]:
    req_obj = global_product_management_service.requirements_traceability.create_requirement(
        product_id=id,
        title=req.title,
        description=req.description,
        category=req.category,
        priority=req.priority,
        acceptance_criteria=req.acceptance_criteria,
        source=req.source,
        problem_id=req.problem_id,
        objective_id=req.objective_id,
    )
    return {"status": "SUCCESS", "data": req_obj}


@router.get("/{id}/backlog", summary="List Product Backlog & Epics")
async def list_backlog(id: str) -> Dict[str, Any]:
    epics = global_product_management_service.backlog_prioritization.list_epics(id)
    features = global_product_management_service.backlog_prioritization.list_features(id)
    items = global_product_management_service.backlog_prioritization.list_backlog_items(id)
    return {"status": "SUCCESS", "epics": epics, "features": features, "items": items}


@router.post("/{id}/backlog/score/{item_id}", summary="Prioritize Backlog Item (RICE/WSJF)")
async def score_backlog_item(id: str, item_id: str, req: PrioritizationScoreRequest) -> Dict[str, Any]:
    score = global_product_management_service.backlog_prioritization.score_item(
        item_id=item_id,
        product_id=id,
        framework=req.framework,
        inputs=req.inputs,
    )
    return {"status": "SUCCESS", "data": score}


@router.get("/{id}/roadmap", summary="Get Roadmaps & Scenarios")
async def get_roadmaps(id: str) -> Dict[str, Any]:
    roadmaps = global_product_management_service.roadmaps_capacity.list_roadmaps(id)
    return {"status": "SUCCESS", "data": roadmaps}


@router.post("/{id}/roadmap", status_code=status.HTTP_201_CREATED, summary="Create Roadmap Scenario")
async def create_roadmap(id: str, req: RoadmapCreateRequest) -> Dict[str, Any]:
    rm = global_product_management_service.roadmaps_capacity.create_roadmap(
        product_id=id,
        title=req.title,
        scenario=req.scenario,
        description=req.description,
    )
    return {"status": "SUCCESS", "data": rm}


@router.get("/{id}/sprints-releases", summary="List Sprints and Releases")
async def list_sprints_releases(id: str) -> Dict[str, Any]:
    sprints = global_product_management_service.sprints_releases.list_sprints(id)
    releases = global_product_management_service.sprints_releases.list_releases(id)
    return {"status": "SUCCESS", "sprints": sprints, "releases": releases}


@router.post("/{id}/releases/{release_id}/readiness", summary="Evaluate Release Readiness Gate")
async def evaluate_readiness(id: str, release_id: str, req: ReleaseReadinessEvaluateRequest) -> Dict[str, Any]:
    gate = global_product_management_service.sprints_releases.evaluate_release_readiness(
        release_id=release_id,
        product_id=id,
        qa_passed=req.qa_passed,
        critical_defects=req.critical_defects,
        security_reviewed=req.security_reviewed,
        performance_benchmarked=req.performance_benchmarked,
        rollback_tested=req.rollback_tested,
    )
    return {"status": "SUCCESS", "data": gate}


@router.get("/{id}/deployments-launches", summary="Get Deployments, Feature Flags & Launch Checklists")
async def get_deployments_launches(id: str) -> Dict[str, Any]:
    flags = global_product_management_service.deployments_launches.list_flags(id)
    launches = global_product_management_service.deployments_launches.list_launches(id)
    return {"status": "SUCCESS", "feature_flags": flags, "launches": launches}


@router.get("/{id}/analytics-experiments", summary="Get Adoption Analytics and A/B Experiments")
async def get_analytics_experiments(id: str) -> Dict[str, Any]:
    adoption = global_product_management_service.analytics_experimentation.get_adoption_overview(id)
    experiments = global_product_management_service.analytics_experimentation.list_experiments(id)
    return {"status": "SUCCESS", "adoption": adoption, "experiments": experiments}


@router.get("/{id}/health", summary="Get Composite Product Health & Risk Factors")
async def get_product_health(id: str) -> Dict[str, Any]:
    health = global_product_management_service.health_sunset.get_latest_health(id)
    history = global_product_management_service.health_sunset.list_health_history(id)
    return {"status": "SUCCESS", "current_health": health, "history": history}


@router.post("/{id}/copilot", summary="Query Product Copilot")
async def query_copilot(id: str, req: ProductCopilotRequest) -> Dict[str, Any]:
    try:
        res = global_product_management_service.query_product_copilot(product_id=id, query=req.query)
        return {"status": "SUCCESS", "data": res}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
